launch/launch/launch_service.py

```diff
@@ -176,6 +174,116 @@ def _is_idle(self):
+
+            def _on_sigint(signum, frame, prev_handler):
+                # Ignore additional signals until we finish processing this one.
+                current_handler = signal.signal(signal.SIGINT, signal.SIG_IGN)
+                if current_handler is signal.SIG_IGN:
+                    # This function has been called re-entrantly.
+                    return
+                nonlocal sigint_received
+                base_msg = 'user interrupted with ctrl-c (SIGINT)'
+                if not sigint_received:
+                    self.__logger.warning(base_msg)
+                    ret = self._shutdown(
+                        reason='ctrl-c (SIGINT)', due_to_sigint=True, force_sync=True
+                    )
+                    assert ret is None, ret
+                    sigint_received = True
+                else:
+                    self.__logger.warning('{} again, ignoring...'.format(base_msg))
+                if callable(prev_handler):
+                    try:
+                        # Run pre-existing signal handler.
+                        prev_handler(signum, frame)
+                    except KeyboardInterrupt:
+                        # Ignore exception.
+                        pass
+                # Restore current signal handler (not necessarily this one).
+                signal.signal(signal.SIGINT, current_handler)
+
+            on_sigint(_on_sigint)
+
+            def _on_sigterm(signum, frame, prev_handler):
+                # Ignore additional signals until we finish processing this one.
+                current_handler = signal.signal(signal.SIGTERM, signal.SIG_IGN)
+                if current_handler is signal.SIG_IGN:
+                    # This function has been called re-entrantly.
+                    return
+                # TODO(wjwwood): try to terminate running subprocesses before exiting.
+                self.__logger.error('using SIGTERM or SIGQUIT can result in orphaned processes')
+                self.__logger.error('make sure no processes launched are still running')
+                this_loop.call_soon(this_task.cancel)
+                if callable(prev_handler):
+                    # Run pre-existing signal handler.
+                    prev_handler(signum, frame)
+                # Restore current signal handler (not necessarily this one).
+                signal.signal(signal.SIGTERM, current_handler)
+
+            on_sigterm(_on_sigterm)
+
+            def _on_sigquit(signum, frame, prev_handler):
+                # Ignore additional signals until we finish processing this one.
+                current_handler = signal.signal(signal.SIGQUIT, signal.SIG_IGN)
+                if current_handler is signal.SIG_IGN:
+                    # This function has been called re-entrantly.
+                    return
+                self.__logger.error('user interrupted with ctrl-\\ (SIGQUIT), terminating...')
+                _on_sigterm(signum, frame, prev_handler)
```

Comment:
Why is `SIGTERM` disabled when `SIGQUIT` is sent?
I know we were doing that before too.

Reply:
That is a good point, though, since the side effect of a SIGQUIT is also deferred termination of the run loop task, ignoring SIGTERM seems innocuous to me. Thoughts?

Reply:
Yes, I agree it's innocuous.